library/pk.c

```diff
@@ -518,6 +523,38 @@ int mbedtls_pk_sign( mbedtls_pk_context *ctx, mbedtls_md_type_t md_alg,
                                          f_rng, p_rng, NULL ) );
 }
 
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+/*
+ * Make a signature with options
+ */
+int mbedtls_pk_sign_ext( mbedtls_pk_type_t pk_type, mbedtls_pk_context *ctx,
+                    mbedtls_md_type_t md_alg,
+                    const unsigned char *hash, size_t hash_len,
+                    unsigned char *sig, size_t sig_size, size_t *sig_len,
+                    int (*f_rng)(void *, unsigned char *, size_t), void *p_rng )
+{
+    const mbedtls_pk_info_t *pk_info = mbedtls_pk_info_from_type( pk_type );
+    PK_VALIDATE_RET( ctx != NULL );
+    PK_VALIDATE_RET( ( md_alg == MBEDTLS_MD_NONE && hash_len == 0 ) ||
+                     hash != NULL );
+    PK_VALIDATE_RET( sig != NULL );
+
+    *sig_len = 0;
+    if( pk_info == NULL )
+        return( MBEDTLS_ERR_PK_BAD_INPUT_DATA );
+
+    if( ! mbedtls_pk_can_do( ctx, pk_type ) )
+        return( MBEDTLS_ERR_PK_TYPE_MISMATCH );
+
+    if( pk_info->sign_func == NULL )
+        return( MBEDTLS_ERR_PK_TYPE_MISMATCH );
+
+    return( pk_info->sign_func( ctx->pk_ctx, md_alg, hash, hash_len,
```

Comment:
I understand it makes sense to use a `pk_info` like the other key types, and it would probably make sense if we wanted to extend PK to support RSA-PSS as a proper key type, but so far it's not really a key type, just a sort of algorithm identifier (unfortunately PK uses the same type for both, it's not as cleanly designed as PSA). And considering the future of PK is uncertain, I'd rather do the minimal amount of changes to it (also to reduce testing burden: I think the new function `rsassa_pss_debug` that you had to introduce to make the `pk_info` complete is untested).

So, I think we should not add a `pk_info` and instead the code structure of `pk_sign_ext()` should be closer to that of `pk_verify_ext()`. Currently the structure of `pk_verify_ext()` is:
```
/* common checks (info != NULL, can_do) */
if( type != MBEDTLS_PK_RSASSA_PSS )
    /* call pk_verify */
/* common stuff with options */
#if defined(MBEDTLS_USE_PSA_CRYPTO)
if( options_are_ok )
    /* do it with PSA */
else
#else
    /* do it with non-PSA */
#endif 
```

If think we can do something similar here but even simpler as we have no options and don't need to have a non-PSA version for compatibility, as this is a new function:
```
/* common checks (info != NULL, can_do) */
if( type != MBEDTLS_PK_RSASSA_PSS )
    /* call pk_verify */
#if defined(MBEDTLS_PSA_CRYPTO_C) // not USE_PSA_CRYPTO
    /* do it with PSA */
#else 
    return FEATURE_UNAVAILABLE
#endif 
```
The "do it with PSA" part may consist of a call to the function `rsa_sign_wrap_ext()` that you introduced (which you may declare in `pk_wrap.h` instead so that it can be used internally), or you can also just directly include similar code here (as it done in `pk_verify_ext()`) - the code will be a bit as we don't need to re-construct a `pk_context_t`, we already have one.

Reply:
`mbedtls_pk_sign_ext` already guards with `MBEDTLS_PSA_CRYPTO_C`.  So I think it should be 
```c
/* common checks (info != NULL, can_do) */
if( type != MBEDTLS_PK_RSASSA_PSS )
{
    /* call pk_verify */
   return;
}

    /* do it with PSA */

``

Reply:
Yes indeed, if we guard the whole function with `PSA_CRYPTO_C` then we can simplify the body. (My suggestion was in case we want to make the function available unconditionally, but that's not useful, so let's go with your solution instead.)

Reply:
CI fail. I rollback to your suggestion.
`psa_algorithm_t` is guards with `MBEDTLS_USE_PSA_CRYPTO` . 

As my understand, `rsa_sign_wrap_ext` should be guards with `PSA_CRYPTO_C` also. Am I wrong? 
But in current code, due to `psa_algorithm_t` , I have to guards it with `MBEDTLS_USE_PSA_CRYPTO`

Is it expected? Does the exists code wrong? an example is in `pk_wrap.c line 51`.